File: tools/verify/render_ocean.c
/* C side for the ocean scene. Prints the same per-frame digests as
 * render_ocean.js, which drives the original JS in Chromium. */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../../core/deck.h"
#include "../../core/screens.h"

#define W 192
#define H 48

static uint32_t fnv1a(const uint8_t *p, size_t n) {
  uint32_t h = 2166136261u;
  for (size_t i = 0; i < n; i++) { h ^= p[i]; h *= 16777619u; }
  return h;
}

int main(int argc, char **argv) {
  const char *path = argc > 1 ? argv[1] : "tools/verify/ocean.tsv";
  FILE *f = fopen(path, "r");
  if (!f) { perror(path); return 1; }

  static uint8_t px[W * H];
  const deck_geom_t geom = { W, H, DECK_LEVELS, 0 };
  deck_fb_t fb = { &geom, px };
  static deck_state_t v;
  static deck_ocean_t o;

  char line[256];
  while (fgets(line, sizeof line, f)) {
    if (line[0] == '/' || line[0] == '\n') continue;
    char name[64];
    int frames;
    double rms, hf, bass, bassStep;
    if (sscanf(line, "%63s %d %lf %lf %lf %lf",
               name, &frames, &rms, &hf, &bass, &bassStep) != 6) continue;

    deck_ocean_reset(&o);
    memset(&v, 0, sizeof v);

    printf("%-14s", name);
    for (int t = 0; t < frames; t++) {
      deck_clear(&fb);
      v.rms01 = rms;
      v.hfAvg = hf;
      v.bassAvg = bass + (t > 8 ? bassStep : 0.0);
      deck_screen_ocean(&fb, &v, &o, (uint32_t)t);

      if (getenv("DECK_DEBUG")) {
        const deck_dolphin_t *d = &o.pod[0];
        printf("\nt%d x%.3f y%.3f m%d jt%d a%.3f f%d sp%d",
               t, d->x, d->ry, d->mode, d->jt, d->rang, d->rflex, o.nSpray);
        continue;
      }
      int nz = 0; long sum = 0;
      for (int i = 0; i < W * H; i++) { if (px[i]) nz++; sum += px[i]; }
      printf(" %d:%x:%d:%ld", t, fnv1a(px, sizeof px), nz, sum);
    }
    printf("\n");
  }
  fclose(f);
  return 0;
}

File: tools/verify/render_ocean.js
// Reference side for the ocean scene.
//
// Unlike every other screen, this one cannot run in bare node: the dolphin
// silhouettes come out of Canvas. So it runs the original dolphin.js in real
// Chromium and prints the same digests render_ocean.c does.
//
// The scene keeps a world between frames, so each case gets a fresh page —
// that is what resets the module-level pod, spray and bubbles.
const fs = require("fs");
const path = require("path");
const { chromium } = require("playwright-core");

const W = 192, H = 48;
const ROOT = path.join(__dirname, "..", "..");
const EXE = process.env.CHROMIUM || "/opt/pw-browsers/chromium-1194/chrome-linux/chrome";

const HARNESS = `
  window.__fb = new Uint8Array(${W} * ${H});
  window.setDot = function (f, x, y, i) {
    if (x < 0 || x >= ${W} || y < 0 || y >= ${H}) return;
    const idx = y * ${W} + x;
    if (i > f[idx]) f[idx] = i;
  };
  window.fnv1a = function (buf) {
    let h = 2166136261 >>> 0;
    for (const b of buf) { h = (h ^ b) >>> 0; h = Math.imul(h, 16777619) >>> 0; }
    return h >>> 0;
  };
`;

(async () => {
  const cases = fs.readFileSync(
    process.argv[2] || path.join(ROOT, "tools", "verify", "ocean.tsv"), "utf8");
  const dolphin = fs.readFileSync(path.join(ROOT, "legacy", "web", "dolphin.js"), "utf8");

  const browser = await chromium.launch({ executablePath: EXE });
  for (const raw of cases.split("\n")) {
    if (!raw || raw.startsWith("/")) continue;
    const [name, framesS, rmsS, hfS, bassS, bassStepS] = raw.trim().split(/\s+/);
    if (!name || framesS === undefined) continue;

    const page = await browser.newPage();          // fresh world per case
    await page.setContent("<canvas id=c></canvas>");
    await page.addScriptTag({ content: HARNESS });
    await page.addScriptTag({ content: dolphin });
    if (process.env.DECK_DEBUG) await page.evaluate(() => { window.__DEBUG = 1; });

    const digest = await page.evaluate((a) => {
      const [frames, rms, hf, bass, bassStep] = a;
      let out = "";
      for (let t = 0; t < frames; t++) {
        window.__fb.fill(0);
        // bass ramps so the breach trigger fires partway through
        const energy = { rms01: rms, hf01: hf, bass01: bass + (t > 8 ? bassStep : 0) };
        oceanFrame(window.__fb, t, energy);
        let nz = 0, sum = 0;
        for (const b of window.__fb) { if (b) nz++; sum += b; }
        if (window.__DEBUG) {
          const d = _pod[0];
          out += `t${t} x${d.x.toFixed(3)} y${(d.render ? d.render.y : 0).toFixed(3)} ` +
                 `m${d.mode} jt${d.jt} a${(d.render ? d.render.ang : 0).toFixed(3)} ` +
                 `f${d.render ? d.render.flex : 0} sp${_spray.length}\n`;
        } else out += `${t}:${fnv1a(window.__fb).toString(16)}:${nz}:${sum} `;
      }
      return out.trim();
    }, [Number(framesS), Number(rmsS), Number(hfS), Number(bassS), Number(bassStepS)]);

    await page.close();
    console.log(`${name.padEnd(14)} ${digest}`);
  }
  await browser.close();
})();
